### ISSUE TITLE
add offset option to time_bucket_gapfill

### ISSUE THREAD

**Issue by mmouterde** [timescale/timescaledb#1304]
**Relevant system information:**
 - TimescaleDB version : 1.3.0

**Describe the bug**
- I'm using `time_bucket` function with the `offset` option. 
- I would like to use the new gapfilling feature but the `offset` option is not provided.
- I can't benefit from the gapfilling feature without regressions.

**Expected behavior**
 `time_bucket_gapfill` takes an `offset` option

**Comment by k-rus** [comment]
Thank you for requesting the feature. We will followup.

**Comment by allstream** [comment]
Origin should be implemented as well, it's hard to use the gapfilling features without origin and offset since the data being aggregated isn't even within the range you essentially query for, but instead is seemingly the closest multiple of the bucket size from the start time and on...

**Comment by ricky-sb** [comment]
>instead is seemingly the closest multiple of the bucket size from the start time and on...  

Could you elaborate what you mean here?   

And does anyone have any workarounds at the moment to implement this feature? Like could you get feature parity by tuning your query parameters and dealing with it on the client?

**Comment by PhilippS93** [comment]
Hi everyone,
is this already on the roadmap? 

Thanks!

**Comment by ateuber** [comment]
I spent too much time today trying to achieve this with the workaround mentioned in #1545, but it is too complicated (we also need to interpolate). I also first expected that the start parameter would be enough, so I think this behavior should be mentioned in [the time_bucket_gapfill documentation](https://docs.timescale.com/latest/api#time_bucket_gapfill).

**Comment by Sytten** [comment]
Can we have this on the roadmap I also spent half a day before realizing why all my days were wrong :(

**Comment by eloyekunle** [comment]
it's a big API inconsistency since one would naturally expect `time_bucket` and `time_bucket_gapfill` to have similar behaviour everywhere else apart from the gap-filling itself.

**Comment by brakonstantin** [comment]
Is there a way to find out whether this issue is going to be addressed?

**Comment by jonatas** [comment]
Hello @brakonstantin, we coordinate our efforts based on our community votes. [Here are](https://timescaledb.slack.com/archives/C4GT3N90X/p1627283364014100)  the latest priorities. Please upvote to help us to prioritize it in the next months.

**Comment by diego-escobedo** [comment]
This would be awesome!

**Comment by tomasrosauer** [comment]
This would be very useful!

**Comment by Rayman-katib** [comment]
This would be awesome!

**Comment by jontingvold** [comment]
I'm also missing this feature.

**Comment by melicheradam** [comment]
+1, Would be great

**Comment by zaycev** [comment]
+1, this would be very handy

**Comment by aabrodskiy** [comment]
is there any update on this issue?
been open since 2019 and without the "origin" parameter the time_bucket_gapfill function is effectively useless as it does not align with the requested datetime period.

**Comment by miguetronic** [comment]
Please make it real!

**Comment by MichaelSaucier** [comment]
> Origin

Absolutely. We need the same fix for offset and origin to make time_bucket_gapfill work.  Without these fixes, we are stuck writing the same function with the fixes, but without the performance and confidence of longevity.

**Comment by DaggerJackfast** [comment]
+1 it would be great.

**Comment by webbegole** [comment]
+1 would make my current projects much easier!

**Comment by benjsec** [comment]
This issue title refers specifically to `offset`, but some comments also mention `origin` and https://github.com/timescale/timescaledb/issues/1545 which was about `origin` has been closed as a duplicate of this. If the two are being considered as the same task can the issue title be updated to help those of us wanting the `origin` parameter, please?

**Comment by jamessewell** [comment]
> This issue title refers specifically to `offset`, but some comments also mention `origin` and [#1545](https://github.com/timescale/timescaledb/issues/1545) which was about `origin` has been closed as a duplicate of this. If the two are being considered as the same task can the issue title be updated to help those of us wanting the `origin` parameter, please?

Added origin to the title 😊

**Comment by MichaelSaucier** [comment]
It has been 5 years and this fundamental and extremely valuable function seems forgotten. As far as "community influence" goes, it has held the number one upvote slot for the last 5 years, with no one from Timescale/TigerData assigned:

<img width="789" height="423" alt="Image" src="https://github.com/user-attachments/assets/a1e25d52-fad1-4e31-9e17-2de6d76df37b" />

There are duplicates (or substantially similar isses) to the original #1304 , like #1435, #3212, and a related but not identical issue #6578.

Please, can we get the attention this bug deserves?  Many thanks!

**Comment by nttoshev** [comment]
Hi all, I have just hit the same problem as the other commenters on this issue - the lack of offset support has become a blocker and I've started implementing hacks around it.

I see a merge requests from January was abandoned, even though several commits were done on it - are there any plans to renew this endeavor?

**Comment by zaycev** [comment]
> Hi all, I have just hit the same problem as the other commenters on this issue - the lack of offset support has become a blocker and I've started implementing hacks around it.
> 
> I see a merge requests from January was abandoned, even though several commits were done on it - are there any plans to renew this endeavor?

It's been 7 years! At this point we already switched to ClickHouse.

**Comment by philkra** [comment]
Thank you for your continued patience and persistence with this issue. This thread has been open for quite a while, and we genuinely appreciate the community continuing to share use cases, workarounds, and upvotes over the years.

We agree with the core feedback here: the API inconsistency between `time_bucket` and `time_bucket_gapfill` regarding the `offset` and `origin` parameters causes unnecessary friction. I want to be transparent that we cannot commit to a specific timeline or release for this feature just yet. Delivering this with the performance and reliability you expect requires careful implementation, and we are currently balancing it alongside other priorities.

Please know that this issue is very much on our product radar and has not been forgotten. Your feedback directly shapes our backlog, and we’ll be sure to update this thread as soon as we have a concrete update on its inclusion in an upcoming release.

**Comment by MichaelSaucier** [comment]
Do timescale folks actually read the upvoted issues table?  This critical defect is unchanged in 7 years.
